### PR TITLE
add min and max replica configuration through crd

### PR DESCRIPTION
**Pull request by OliverMKing**
# Description

lets users configure min and max replicas through crd

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Unit, e2e, and manually tested.

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: pkg/manifests/nginx_test.go [hunk 1]
```diff
@@ -22,6 +22,8 @@ var (
 				"service.beta.kubernetes.io/azure-load-balancer-internal": "true",
 			},
 		},
+		MinReplicas: 2,
```
Comment (jaiveerk) [inline pkg/manifests/nginx_test.go:25]: can you add a fixture test for when both scale values are nil? That's a possible case, right?

**Review by jaiveerk — CHANGES_REQUESTED** [review]
lgtm just a couple of test requests

File: config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml [hunk 2]
```diff
@@ -104,6 +104,26 @@ spec:
...
+                    type: integer
+                  minReplicas:
+                    description: MinReplicas is the lower limit for the number of
+                      Ingress Controller replicas. It defaults to 2 pods.
+                    format: int32
+                    minimum: 1
+                    type: integer
+                type: object
+                x-kubernetes-validations:
+                - rule: (!has(self.minReplicas)) || (!has(self.maxReplicas)) || (self.minReplicas
```
Comment (bosesuneha) [inline config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml:125]: Just a question since I'm not familiar with this: why does the rule check if minReplicas or maxReplicas are not set? Also, is there any limit for maxReplica count?

File: pkg/controller/nginxingress/nginx_ingress_controller_test.go [hunk 3]
```diff
@@ -954,6 +966,145 @@ func TestToNginxIngressConfig(t *testing.T) {
...
+				ControllerClass: "approuting.kubernetes.azure.com/nicName",
+				ResourceName:    "controllerNamePrefix-0",
+				ServiceConfig:   &manifests.ServiceConfig{},
+				IcName:          "ingressClassName",
+				MaxReplicas:     47,
+				MinReplicas:     defaultMinReplicas,
+			},
+		},
+		{
+			name: "custom fields with max replica lower than min",
```
[the remark this answers is not in the document]
Comment (OliverMKing) [reply pkg/controller/nginxingress/nginx_ingress_controller_test.go:1064]: that's covered by cel validation in the crd. we purposefully left it off here but test it in e2e

File: pkg/manifests/nginx_test.go (hunk 1 above)
Comment (OliverMKing) [reply pkg/manifests/nginx_test.go:25]: that is not. the options struct used to generate manifest doesn't use a pointer.

the nil is transformed at a higher level where this is fully tested

File: config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml (hunk 2 above)
Comment (OliverMKing) [reply config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml:125]: they are pointer values so it's possible for a user to not have them set or only one to be set. in CEL (the validation language) there's no default values so if we were to compare something null vs a defined thing it would panic.

there is no limit for max replica count. this matches the form of the hpa which this field eventually flows into

File: pkg/controller/nginxingress/nginx_ingress_controller.go [hunk 4]
```diff
@@ -30,6 +30,11 @@ const (
 	controllerClassMaxLen = 250
 )
 
+const (
+	defaultMinReplicas = 2
+	defaultMaxReplicas = 100
+)
```
Comment (Vidya2606) [inline pkg/controller/nginxingress/nginx_ingress_controller.go:36]: Question: on what basis are we setting the value for min and max replicas?

Comment (OliverMKing) [reply pkg/controller/nginxingress/nginx_ingress_controller.go:36]: this is the default. It's what we previously had which is important, so the behavior doesn't change.

we have scale tested these and these are acceptable defaults

File: pkg/manifests/nginx_test.go (hunk 1 above)
Comment (jaiveerk) [reply pkg/manifests/nginx_test.go:25]: got it - and that's in the controller tests right?

File: config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml (hunk 2 above)
Comment (jaiveerk) [reply config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml:125]: if customers want scale... we'll give em scale! 😆

**Review by jaiveerk — APPROVED** [review]
lgtm

**Review by Vidya2606 — DISMISSED** [review]

File: pkg/manifests/nginx_test.go (hunk 1 above)
Comment (OliverMKing) [reply pkg/manifests/nginx_test.go:25]: exactly

File: config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml (hunk 2 above)
Comment (OliverMKing) [reply config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml:125]: also note that this flows down to an HPA which will only scale to the max if that number is actually needed

Comment (bosesuneha) [reply config/crd/bases/approuting.kubernetes.azure.com_nginxingresscontrollers.yaml:125]: thanks, that makes sense!
